msrestazure/azure_local_creds_prober.py

```diff
@@ -0,0 +1,280 @@
+        return creds
+
+    def probe_subscription(self, creds):
+        creds.set_token()
+        return creds.subscription_id
+
+
+class CLICredentials(BasicTokenAuthentication):
+    '''
+    Credentials objects for AAD token retrieved through sub-shelling "az account get-access-token"
+    '''
+    def __init__(self, cli_path, subscription_id=None):
+        super(CLICredentials, self).__init__(None)
+        self.cli_path = cli_path
+        self.subscription_id = subscription_id
+        self.expires_on = None
+        self.resource = None
+        self.cloud_environment = None
+
+    def set_token(self):
+        from dateutil import parser
+        from datetime import timedelta, datetime
+        if (not self.token or self.expires_on and
+                (datetime.now() + timedelta(minutes=5)) > self.expires_on):
+            info = self._invoke_cli_token_command()
+            self.scheme, self.token, self.expires_on = (info['tokenType'],
+                                                        {'access_token': info['accessToken']},
+                                                        parser.parse(info['expiresOn']))
+            if self.subscription_id is None:
+                self.subscription_id = info['subscription']
+        return self.scheme, self.token
+
+    def _invoke_cli_token_command(self):
+        args = [self.cli_path, 'account', 'get-access-token']
+        if self.subscription_id:
+            args.extend(['--subscription', self.subscription_id])
+        if self.resource:
+            args.extend(['--resource', self.resource])
+        process = Popen(args, stdout=PIPE, stderr=PIPE)
+        stdout, stderr = process.communicate()
+        process.wait()
+        if stderr:
+            raise ValueError('Retrieving acccess token failed: ' + stderr)
+        return json.loads(stdout.decode())
+
+    def signed_session(self, session=None):
+        self.set_token()
+        return super(CLICredentials, self).signed_session(session)
+
+
+def _get_creds_through_local_probing(**kwargs):
+    resource = kwargs.get('resource')
+    probers = [ConnectionStrEnvProber(resource), ServicePrincipalEnvProber(resource),
+               ManagedServiceIdentityProber(resource), AzureCLIProber(resource)]
+    for prober in probers:
+        creds = prober.probe()
+        if creds:
+            return creds, prober
+    raise ValueError('No credential was detected from the local machine')
+
+
+def get_creds_through_local_probing(**kwargs):
+    '''
+    Probing logics:
+    1. AZURE_CONN_STR, with SDK auth code file content in json.
+        https://github.com/Azure/azure-sdk-for-java/wiki/Authentication
+    2. Individual environment variables to estabslish a service principal's creds
+        https://github.com/Azure/azure-sdk-for-go
+    3. Managed service identity
+        a. app service
+        b. virtual machine
+    4. Azure CLI, through "az account get-access-token"
+    '''
```

Comment:
Add the authentication file:
https://docs.microsoft.com/en-us/python/azure/python-sdk-azure-authenticate?view=azure-python#mgmt-auth-file